Add stack file template configuration to build

martindekov opened the pull request:
Adding functionality which will pull templates defined in the
function YAML file. Additionaly added overwrite flag to the
build command in order to enable the command to overwrite
templates. Also removed the filtering funcitonality from template
pull stack command and renamed the method which will be used to
pull the configuration

Signed-off-by: Martin Dekov <mvdekov@gmail.com>

<!--- Provide a general summary of your changes in the Title above -->

Configuration in stack.yml file was used only by `template pull stack` command. Now the configuration is additionally red in the building phase of the function.

## Description
<!--- Describe your changes in detail -->

Adding `--overwrite` flag to the `faas-cli build` command as it reads and pull templates
defined in the function YAML file. Removing the filtering logic of the `template pull stack`
along with its testing as requested and renaming the method responsible for the pulling. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

Closes #725 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Appended to a new function's `stack.yml` the following:

```
configuration:
  templates:
    - name: perl-alpine
    - name: rusta
      source: https://github.com/openfaas-incubator/openfaas-rust-template
```
and tested
* `faas-cli build` - pulled templates if not there, if there used them cached
* `faas-cli build --overwrite` - pulled templates no matter present or not
* `faas-cli template pull stack` - same as above
* `faas-cli template pull stack --overwrite` - same as above

Additional case when there are already existing templates(`perl-alpine`,`rust`,`valum-http`) and we decide to add additional template(`crystal-http`) to the configuration the following happens.

```
$ faas-cli build 
Pulling template: perl-alpine from configuration file: stack.yml
Fetch templates from repository: https://github.com/tmiklas/openfaas-perl-templates at master
2019/11/16 13:21:32 Attempting to expand templates from https://github.com/tmiklas/openfaas-perl-templates
2019/11/16 13:21:34 Cannot overwrite the following 1 template(s): [perl-alpine]
2019/11/16 13:21:34 Fetched 0 template(s) : [] from https://github.com/tmiklas/openfaas-perl-templates
Pulling template: rusta from configuration file: stack.yml
Fetch templates from repository: https://github.com/openfaas-incubator/openfaas-rust-template at master
2019/11/16 13:21:34 Attempting to expand templates from https://github.com/openfaas-incubator/openfaas-rust-template
2019/11/16 13:21:36 Cannot overwrite the following 1 template(s): [rust]
2019/11/16 13:21:36 Fetched 0 template(s) : [] from https://github.com/openfaas-incubator/openfaas-rust-template
Pulling template: vala-http from configuration file: stack.yml
Fetch templates from repository: https://github.com/affix/openfaas-templates-affix/ at master
2019/11/16 13:21:36 Attempting to expand templates from https://github.com/affix/openfaas-templates-affix/
2019/11/16 13:21:37 Cannot overwrite the following 4 template(s): [lua53 swift vala-valum vala-valum-http]
2019/11/16 13:21:37 Fetched 0 template(s) : [] from https://github.com/affix/openfaas-templates-affix/
Pulling template: crystal-http from configuration file: stack.yml
Fetch templates from repository: https://github.com/koffeinfrei/crystal-http-template at master
2019/11/16 13:21:37 Attempting to expand templates from https://github.com/koffeinfrei/crystal-http-template
2019/11/16 13:21:39 Fetched 1 template(s) : [crystal-http] from https://github.com/koffeinfrei/crystal-http-template
```
Removed the appended template configuration and re-ran the tests mentioned above and the building process was as it were without the change.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

rgee0 commented:
I'm not convinced about the UX here.  I know the issue said pull the templates only if the template folder doesn't exist but lets consider where the template folder *does* exist but the *language template* isn't contained within.  Should this work without overwrite?  As its non-destructive it feels like it ought to.

martindekov commented:
Hey Richard, I can remove overwrite flag from the build command and in case you want to overwrite something you can be explicit and use `template pull stack --overwrite`. Would that be better? WDYT?

rgee0 commented:
Possibly (having an `--overwrite` flag on `build` does seem a little unintuitive).  I think what i was thinking more of was transparently pulling the stack's templates on build in scenarios where the stack's language folder(s) don't already exist.  So, as an example, we could have a situation where the standard templates are present and the stack is using two non-standard templates, here `faas build` should not clobber the existing templates and should successfully build without the user having to provide the `--overwrite` flag

LucasRoesler commented on commands/template_pull_stack.go:
```diff
@@ -98,14 +94,3 @@ func findTemplate(templateInfo []stack.TemplateSource, customName string) (speci
 	return nil
 }
-
-func pullSpecificTemplate(templateInfo []stack.TemplateSource, customName string, cmd *cobra.Command) error {
```
Should this functionality really be removed? How should this behave when `regex` or `filter` is passed to the build?   These are persistent flags that are supposed to match function names in the yaml file

https://github.com/openfaas/faas-cli/blob/83081ef8ff0dbc2c8fb53954188a0f0f1f7e8e0f/commands/faas.go#L63-L64

Should the pull stack then ensure that it pulls the templates required for the matching functions?  What is the template folder exists but is missing a new template for a function that was appended to the file.  Perhaps pulling a single template is still a good idea

martindekov replied:
Hey Lucas, interesting questions.

The first cluster of questions, the regex and filter are just skipped for the template pull stack. We handle configuration field only for the new functionality. Those flags are not handled in `pull stack` so they will have no affect on pulling.

For the second wave it will just pull the missing template. So if I have 3 templates and add 4th, try pulling, the 3 templates will be cached and won't be pulled, but the 4th one will be pulled as it does not exist.

I synced with Alex about this, he requested it in order for the feature to be more simple/elegant. @alexellis WDYT? Should we leave the feature, or remove it?

martindekov replied:
So I ran this test
```
...
  newfuncyaw:
    lang: perl-alpine
    handler: ./newfuncyaw
    image: martindekov/newfuncyaw:latest
  saws:
    lang: python
    handler: ./saws
    image: martindekov/saws:latest

configuration:
  templates:
    - name: rusta
      source: https://github.com/openfaas-incubator/openfaas-rust-template
    - name: vala-http
    - name: crystal-http
```
`faas-cli build --regex="saws"`, the function build sucessfully, even though the first function did not have the `perl-alpine` template pulled(it would have got error.):
```
sudo ../../openfaas/faas-cli/faas-cli build --regex="saws"
Pulling template: rusta from configuration file: stack.yml
Fetch templates from repository: https://github.com/openfaas-incubator/openfaas-rust-template at master
2019/11/16 13:32:12 Attempting to expand templates from https://github.com/openfaas-incubator/openfaas-rust-template
2019/11/16 13:32:13 Cannot overwrite the following 1 template(s): [rust]
2019/11/16 13:32:13 Fetched 0 template(s) : [] from https://github.com/openfaas-incubator/openfaas-rust-template
Pulling template: vala-http from configuration file: stack.yml
Fetch templates from repository: https://github.com/affix/openfaas-templates-affix/ at master
2019/11/16 13:32:13 Attempting to expand templates from https://github.com/affix/openfaas-templates-affix/
2019/11/16 13:32:15 Cannot overwrite the following 4 template(s): [lua53 swift vala-valum vala-valum-http]
2019/11/16 13:32:15 Fetched 0 template(s) : [] from https://github.com/affix/openfaas-templates-affix/
Pulling template: crystal-http from configuration file: stack.yml
Fetch templates from repository: https://github.com/koffeinfrei/crystal-http-template at master
2019/11/16 13:32:15 Attempting to expand templates from https://github.com/koffeinfrei/crystal-http-template
2019/11/16 13:32:16 Cannot overwrite the following 1 template(s): [crystal-http]
2019/11/16 13:32:16 Fetched 0 template(s) : [] from https://github.com/koffeinfrei/crystal-http-template
[0] > Building saws.
Clearing temporary build folder: ./build/saws/
Preparing ./saws/ ./build/saws//function
Building: martindekov/saws:latest with python template. Please wait..
Sending build context to Docker daemon  8.192kB
Step 1/29 : FROM openfaas/classic-watchdog:0.18.1 as watchdog
 ---> 94b5e0bef891
Step 2/29 : FROM python:2.7-alpine
 ---> 1bf48bb21060
Step 3/29 : ARG ADDITIONAL_PACKAGE
 ---> Using cache
 ---> b683fd89484d
Step 4/29 : COPY --from=watchdog /fwatchdog /usr/bin/fwatchdog
...
```
Hope I got your right this time? The flow of the program can be traced, the building starts after pulling the templates, and the `regex` command will just make it so the function is skipped(if template for the skipped function is missing it will just not start building it, so there would be no error and the regexed function will be built with no problems).

martindekov commented:
I agree with you @rgee0 having overwrite on build is kind of confusing. I will probably remove the flag for this command. Otherwise if the template folder **exists** and we add another template to the configuration **only the new templates will be pulled**. I believe Lucas has the same question in his mind. And following the logic it works without overwrite :+1: I apologize for misreading the question first time, I had a ton of things going around at that moment.

martindekov commented:
Removed `--overwrite` and updated the examples to show that if we add additional template it will just be pulled additionally, the rest of the templates will not be overwritten.

martindekov commented:
Not familiar with `common` folder. Is there an Issue? I am curious, also I think we can stick with the configuration in `stack.yml` saw there is update in `yaml.v3` the most interesting one being the [better handling of comments](https://github.com/go-yaml/yaml/commit/0b070bb63a1840d358e283258faa3b4022ed0396) I need to test it out though. Possibly lookin forward [save stack](https://github.com/openfaas/faas-cli/issues/684).

LucasRoesler commented:
@martindekov feel free to ignore my comment about common folder, it is a separate feature, but the configuration will go inside StackConfiguration

rgee0 commented:
#322 is the common folder discussion

alexellis commented:
@martindekov a rebase is required for this PR. 

Could you also add a flag that can be used by OpenFaaS Cloud and git-tar to turn off the custom template behaviour? i.e. `--disable-stack-pull`?

martindekov commented:
Yeah Alex no problems :+1:

martindekov commented:
Done :+1:

Tested `stack.yml` with content:
```
...
    image: martindekov/saws:latest

configuration:
  templates:
    - name: rusta
      source: https://github.com/openfaas-incubator/openfaas-rust-template
    - name: vala-http
    - name: crystal-http
```
```
$ faas-cli build
Pulling template: rusta from configuration file: stack.yml
Fetch templates from repository: https://github.com/openfaas-incubator/openfaas-rust-template at master
2019/12/10 20:58:22 Attempting to expand templates from https://github.com/openfaas-incubator/openfaas-rust-template
2019/12/10 20:58:24 Cannot overwrite the following 1 template(s): [rust]
2019/12/10 20:58:24 Fetched 0 template(s) : [] from https://github.com/openfaas-incubator/openfaas-rust-template
Pulling template: vala-http from configuration file: stack.yml
Fetch templates from repository: https://github.com/affix/openfaas-templates-affix/ at master
2019/12/10 20:58:25 Attempting to expand templates from https://github.com/affix/openfaas-templates-affix/
2019/12/10 20:58:26 Cannot overwrite the following 4 template(s): [lua53 swift vala-valum vala-valum-http]
2019/12/10 20:58:26 Fetched 0 template(s) : [] from https://github.com/affix/openfaas-templates-affix/
Pulling template: crystal-http from configuration file: stack.yml
Fetch templates from repository: https://github.com/koffeinfrei/crystal-http-template at master
2019/12/10 20:58:26 Attempting to expand templates from https://github.com/koffeinfrei/crystal-http-template
2019/12/10 20:58:28 Cannot overwrite the following 1 template(s): [crystal-http]
2019/12/10 20:58:28 Fetched 0 template(s) : [] from https://github.com/koffeinfrei/crystal-http-template
[0] > Building newfuncyaw.
[0] < Building newfuncyaw done in 0.00s.
[0] > Building saws.
Clearing temporary build folder: ./build/saws/
Preparing ./saws/ ./build/saws//function
Building: martindekov/saws:latest with python template. Please wait..
...
```

And with flag `--disable-stack-pull` present:
```
$ faas-cli build --disable-stack-pull
[0] > Building newfuncyaw.
[0] < Building newfuncyaw done in 0.00s.
[0] > Building saws.
Clearing temporary build folder: ./build/saws/
Preparing ./saws/ ./build/saws//function
Building: martindekov/saws:latest with python template. Please wait..
Sending build context to Docker daemon  8.192kB
...
```
